### ISSUE TITLE
Report System Backend

### ISSUE THREAD

**Issue by LarsFlieger** [dopeshot/truth-or-dare-backend#71]

**Comment by LarsFlieger** [comment]
Wie ist das aktuelle Design der Reports? Das sollten wir ggf. nochmal anpassen an aktuelles Feedback.

**Comment by LarsFlieger** [comment]
> Design im Sinne von API Design?

Ja genau... Gesamt das Feature Design auch fürs Frontend ;)

**Comment by coffemakingtoaster** [comment]
Eventuell wäre es vorteilhaft, wenn @Branchverse hier nochmal kurz seine Meinung dazugibt

**Comment by coffemakingtoaster** [comment]
Sobald wir das fachliche scope geklärt haben, werd ich so schnell wie möglich eine technische Spezifikation bereitstellen

**Comment by LarsFlieger** [comment]
> @Zoe-Bot @LarsFlieger 
> Die Aufteilung in inhaltiches und böswillig halte ich für Sinn befreit.
> Ein Report ist zum Melden von anstößigem Content gedacht, nicht als Weg dem Set creator Feedback zu geben. 
> Da überschneidet sich ein (in weiter Zukunft liegendes und noch nicht ausspezifiziertes) Bewertungssystem (+ evtl. Kommentarfunktion) mit dem Report feature.
> Abgesehen davon sorgt eine Aufteilung in 'für den creator' und 'für admins' bei den Reports dazu, dass ein User, der das Interface (beispielsweise Aufgrund Alkoholeinflusses oder unzureichender Aufmerksamkeit) nicht korrekt verwendet eventuell wichtige Information an den Set creator anstatt an Admins schickt.
> 
> tldr: Reports sollten kein Verbesserungsformular an den Content creator sein.
> 
> Entsprechend wäre mein Verständnis eines Reports:
> Er wird genutzt um anstößige, feindselige, ... Inhalte zu melden. Diese werden ab einem gewissen Threshhold an Reports (vorerst wird das einfach eine Anzahl X sein) den Usern nicht mehr angezeigt, bis ein Admin sich des reports angenommen hat. 
> 

Verstehe deinen Punkt! Reports ist da vielleicht der falsche Begriff für. Eher vielleicht sowas wie "Ein Problem melden". Bei mehreren Runden spielen ist mir aufgefallen, das immer mal wieder irgendwie / irgendwo ein Fehler in den Tasks ist. Um das also in den Griff zu kriegen brauchen wir in irgendeiner Form ein Bewertungssystem bzw. einen Button um das dem Author mitzuteilen. 

Daher habe ich zwei Sachen gemacht: 

1. Mir überlegt wann so ein Problem auftritt. Wie der User darauf reagiert. 
3. Andere Apps angeschaut wie die das lösen

Ergebnis: Reports oder Feedback sind nicht immer voneinander getrennt. Müssen es ja auch gar nicht. Warum soll es in der UI zwei Buttons geben? Anstatt einfach "Ein Problem melden", "Feedback" oder "Report". Dem Enduser ist das völlig egal. Er möchte etwas über die Task / ein Set sagen. So macht es zum Beispiel Google Maps. 

Wenn wir das jetzt auftrennen, dann sehe ich keine Bedarf für reine Reports da wir zu wenig Content und User haben. Erstellt jetzt jemand ein öffentliches Set merken wir das Problem bevor jemand anderes uns über die Report funktion darüber aufmerksam macht. Und schonmal für die Zukunf entwickeln würde ich auch nicht weil wir dafür eigentlich genug zu tun haben für jetzt. 

Das "Bewertungssystem" wäre aber durchaus praktisch weil eigentlich in jedem Set beim spielen noch Fehler gefunden wurden. Klar, das eillt eigentlich auch nicht weil das uns ja wo anders aufschreiben könnten bzw. einfach direkt fixen.

**Comment by coffemakingtoaster** [comment]
@LarsFlieger 
Ich sehe was du meinst 🤔 
Aber das Ding ist, dass du ja selbst sagst, dass man hier zwei Features ineinander baut, wenn man das weiterhin reports nennt. 

>   So macht es zum Beispiel Google Maps.

Google Maps halt ich für ein Semi treffendes Beispiel, weil der user generated content da etwas anders funktioniert als hier und google maps als Informationsquelle dient und nicht als Spiel/Unterhaltung.
Da treffendere Beispiel wäre da Youtube, da das auch Content ist, den ich konsumiere (ob ich den anschaue oder an nem Tisch meinen Freunden vorlese schenkt sich ja erstmal nicht viel). Reportest du Videos, die beispielweise rassistische Inhalte haben und wenn die oft genug reported werden müssen die händisch reviewed werden (teilweise sind die sogar nicht mehr sichtbar, bis sie menschlich geprüft wurden).

Sprich sollten Reports in unserem Falle eine Möglichkeit sein, eine Frage offline zu nehmen wenn sie anstößige Inhalte hat. Dass uns das (auch bei der überschaubaren Anzahl an content) nicht immer gelingt das selbst zuerst zu finden hat die Vergangenheit gezeigt 😅 

Zudem würde das jetzt schon einführen einer Feedback logik vorraussetzten, dass der Set creator dieses Feedback geordnet und sinnvoll dargstellt bekommt, was meiner Meinung nach den Rahmen von v.0.2 sprengt.

> Das "Bewertungssystem" wäre aber durchaus praktisch weil eigentlich in jedem Set beim spielen noch Fehler gefunden wurden

Stimmt absolut, aber das  sind schlussendlich zwei unterschiedliche Funktionen (allein von der Zielgrupper her, da reports zu admins und feedback zum creator geht). Entsprechend würde ich behaupten, dass das unterschiedliche Endpunkte und DB Objekte sein sollte -> sollten das dann auch mind. 2 issues sein und nicht eine

**Comment by LarsFlieger** [comment]
> @LarsFlieger Ich sehe was du meinst 🤔 Aber das Ding ist, dass du ja selbst sagst, dass man hier zwei Features ineinander baut, wenn man das weiterhin reports nennt.
> 
> > So macht es zum Beispiel Google Maps.
> 
> Google Maps halt ich für ein Semi treffendes Beispiel, weil der user generated content da etwas anders funktioniert als hier und google maps als Informationsquelle dient und nicht als Spiel/Unterhaltung. Da treffendere Beispiel wäre da Youtube, da das auch Content ist, den ich konsumiere (ob ich den anschaue oder an nem Tisch meinen Freunden vorlese schenkt sich ja erstmal nicht viel). Reportest du Videos, die beispielweise rassistische Inhalte haben und wenn die oft genug reported werden müssen die händisch reviewed werden (teilweise sind die sogar nicht mehr sichtbar, bis sie menschlich geprüft wurden).
> 
> Sprich sollten Reports in unserem Falle eine Möglichkeit sein, eine Frage offline zu nehmen wenn sie anstößige Inhalte hat. Dass uns das (auch bei der überschaubaren Anzahl an content) nicht immer gelingt das selbst zuerst zu finden hat die Vergangenheit gezeigt 😅
> 
> Zudem würde das jetzt schon einführen einer Feedback logik vorraussetzten, dass der Set creator dieses Feedback geordnet und sinnvoll dargstellt bekommt, was meiner Meinung nach den Rahmen von v.0.2 sprengt.
> 
> > Das "Bewertungssystem" wäre aber durchaus praktisch weil eigentlich in jedem Set beim spielen noch Fehler gefunden wurden
> 
> Stimmt absolut, aber das sind schlussendlich zwei unterschiedliche Funktionen (allein von der Zielgrupper her, da reports zu admins und feedback zum creator geht). Entsprechend würde ich behaupten, dass das unterschiedliche Endpunkte und DB Objekte sein sollte -> sollten das dann auch mind. 2 issues sein und nicht eine

Ja, das finde ich auch gut! Eher zwei Issues draus machen. Vorallem sollten Reports auch schneller für uns sichtbar sein, weil sie kritischer sind als Feedback. 

Bei den Reports: Denkst du nicht es ist besser, bei Reports sie nicht automatisch zu hiden? Denke das brauchen wir nicht. 

1. Es ist damit ein Angriffvektor weil User den Content der Website ändern können
2. So viele Reports wird es nicht geben (nach meiner Erfahrung her) 
3. Es sollte immer ein Admin entscheiden anstatt die User selbst

Kennst du ne Website die das macht mit hiden? Ich kenne es nur bei upvotes / downvotes das z.B. der Kommentar nach unten rutscht oder markiert wird wenn er sehr viele downvotes hat. 

Also zusammenfassend haben wir jetzt folgende Optionen:

1. Reports und Feedback zusammen machen (Das schließen wir aus)
2. Reports und Feedback trennen (Das machen wir) 

Das heißt, wir machen jetzt für den Release: 

1. Wir implementieren die Reports
2. Wir implementieren die Feedback
3. Wir implementieren beides
4. Wir implementieren nichts für den Release

Habe ich was vergessen? Was denkst du ist am sinnigsten?

**Comment by coffemakingtoaster** [comment]
Ich halte Option 1 für die sinnvollste. Entsprechend hätte ich gesagt dass die Reports (in simpler Form) implementiert werden.  

Bezüglich Reports als möglicher Angriffsvektor gebe ich dir Recht..entsprechend würde ich das rausnehmen und durch eine Notification Logik ersetzen.

tldr: Nur Reports implementieren und automatische Content suspension rausnehmen.

Falls das so für dich passt würde ich heute Abend die technische Spezifikation (DTOs, Endpunkte,...) definieren. Allerdings hierbei mit Fokus auf die Web app (sprich dem erstellen von Reports)...komplexere Admin facing endpoints würde ich erst in Angriff nehmen wenn das Admin frontend aktiv in Entwicklung ist.
Sprich Endpunkte, die ich vor dem Hintergrund in Angriff nehmen würde:
- POST endpoint für clients
- GET Endpoint für admins 
- PATCH Endpoint for admins (nur für Statuschanges `needing review` => 'reviewed')
@LarsFlieger Bitte um Feedback

**Comment by LarsFlieger** [comment]
Finde ich klingt gut! Lass es uns so machen :) Danke!

**Comment by LarsFlieger** [comment]
Out of context ein bisschen aber aus einer anderen Wahrheit oder Pflicht App:

![Screenshot_20220920-203957.jpg](https://user-images.githubusercontent.com/47084064/191338298-dcddabc5-5b89-4861-9a3d-a8b6d27733ee.jpg)

Wir sollten einen Issue dazu machen :)

**Comment by coffemakingtoaster** [comment]
@LarsFlieger @Zoe-Bot
Api Spec:

## General design choice
3 DB entities:
    - setReport
    - taskReport
    - userReport

3 different endpoints:
    - /reports/set
    - /reports/task
    - /report/user

Separating the possible reports allows for more possibilities when extending this feature.
It also removes the need for convoluted and cluttered entities and DTOs

## POST endpoints
Create new reports for either a task,set or user report
=> All POST endpoints are public without the need for login
POST /reports/set
{
    "setId": mongoId (required),
    "description": string (optional),
    "createdBy": mongoId of reporting user (optional),
    "setOwner": mongoId of set owning user (required),
    "reportReason": string from enum of Content report reasons (required)
}

POST /reports/task
{
    "setId": mongoId (required),
    "taskId": mongoId (required)
    "description": string (optional),
    "createdBy": mongoId of reporting user (optional),
    "setOwner": mongoId of set owning user (required),
    "reportReason": string from enum of Content report reasons (required)
}

POST /reports/user
{
    "userId": mongoId (required),
    "description": string (optional),
    "createdBy": mongoId of reporting user (optional),
    "reportReason": string from  enum of reasons (required)
}

Possible responses:
201 - Created
400 - duh
405 - cannot report items that are owned by you/ cant report yourself
409 - Item has already been reported by user

## PATCH
=> admin only
PATCH /reports/set
PATCH /reports/task
PATCH /reports/user
{
    status: string (based on enum)
}
200 - patched
400 - invalid 
405 - Cannot change report status when status is 'redundant'

## GET
=> admin only
GET /reports/set
{
    "setId": mongoId (required),
    "description": string (optional),
    "createdBy": reporting user (optional),
    "setOwner": owning user (required),
    "status": Status,
    "createdAt": date,
    "reportReason": string from  enum of Content report reasons
}

GET /reports/task
{
    "setId": mongoId,
    "taskId": mongoId,
    "description": string (optional),
    "createdBy": reporting user (optional),
    "setOwner": owning user ,
    "status": Status,
    "createdAt": date,
    "reportReason": string from  enum of Content report reasons
}

GET /reports/user
{
    "userId": mongoId (required),
    "description": string (optional),
    "createdBy": reporting user (optional),
    "status": Status,
    "createdAt": date,
    "reportReason": string from  enum of reasons
}

**Possible Report status**
'awaiting_review' = Report has been submitted and is yet to be reviewed
'admin_approved_report' = Report has been reviewed and the reviewer has agreed that there is an issue
'admin_denied_report' = Report has been reviewed, however the reviewer did not find any issues
'redundant' = Report is redundant as the reported Set/Task/User is no longer in existance

**Possible report reasons**
'harrassment' = Generally everything racist, sexist, etc.
'inappropriate' = Especially meant for questions or set that are not fit for the context (for instance age rating of the set)
'other' = task is just a link/advertisment,... => this includes spam


**User report reasons**
'offensive_username' = name says it all
'questionable_behaviour' = account that for instance only posts racist sets, etc.

**Comment by LarsFlieger** [comment]
@coffemakingtoaster Hab drüber geschaut und mit "⚠️ MC:" mal kommentiert. Finde vor allem die Idee reports aufzutrennen sehr gut! 👍🏻

**Comment by coffemakingtoaster** [comment]
@LarsFlieger 

> Den setOwner hast du hinzugefügt weil das schwierig ist das zu fetchen aus der setId, richtig? Kriegen wir das irgendwie hin ohne? Weil wenn wir es so machen, können die Daten falsch sein indem man eine falsche owner id mit gibt. Haben wir noch eine andere Option?

valide...nehm ich aus dem DTO raus, kann im Backend ausgefüllt werden

>  [...] Reported wurde gespeichert weil User den ja theoretisch ändern könnten...

Ob content geändert wurde lässt sich anhand des createdAt datums des reports und des updatedAt datums von Set/task/user feststellen...den reporteten content speichern halte ich dahingehend für sinnbefreit, wenn sich der content schon geändert hat und nun unbedenklich ist, dann ist die Information unbedenklich. Bei Spam problemen wird die Anzahl an reports ein guter inidcator sein falls der durch ändern des contents versteckt wird...aber wichtiger edge case

> Wenn eine Task mehrfach reportet wurde, zählt das dann auch unter dem Punkt?

Nein in dem Fall nicht. Redundate reports hätte ich als reports bezeichnet, die nicht mehr relevant sind, da der reportete content eben nicht mehr existiert.
Mehrere reports für den gleichen content können (beispielsweise bei fremdsprachigen Sets) für admins von wert sein.
Der Name ist aber glaube ich schlecht gewählt...ich hätte den wert also in `content_deleted` umbenannt um Verwirrung vorzubeugen.

Scheint mir aber so, als könnte die Implementation beginnen, richtig?

**Comment by LarsFlieger** [comment]
> Ob content geändert wurde lässt sich anhand des createdAt datums des reports und des updatedAt datums von Set/task/user feststellen...den reporteten content speichern halte ich dahingehend für sinnbefreit, wenn sich der content schon geändert hat und nun unbedenklich ist, dann ist die Information unbedenklich. Bei Spam problemen wird die Anzahl an reports ein guter inidcator sein falls der durch ändern des contents versteckt wird...aber wichtiger edge case
Ja, damit verlieren wir halt die Information und der User könnte wenn er das System kennt Spam produzieren und später rückgängig machen. Aber das ist ein sehr edge case. Bin da beim Gedanke ein bisschen bei "event sourcing"... aber lass es uns weglassen. :) 

Ja, let's go with coding! :) Scheint alles geklärt zu sein!
